File: Recursion6/MergeSort.py
def sort(arr, start, mid, end):
	i = start
	j = mid+1
	result = []

	while i <= mid and j <= end:
		if arr[i] <= arr[j]:
			result.append(arr[i])
			i += 1
		else:
			result.append(arr[j])
			j += 1
	while i <= mid:
		result.append(arr[i])
		i += 1
	while j <= end:
		result.append(arr[j])
		j += 1

	for k in range(len(result)):
		arr[start + k] = result[k]

	return arr

	for k in range(len(arr)):
		arr[start + k] = result[k]



def mergeSort(arr, start, end):
	if start >= end: return arr

	mid = start + (end - start) // 2
	mergeSort(arr, start, mid)
	mergeSort(arr, mid + 1, end)
	sort(arr, start, mid, end)

	return arr
	



arr = [5,8,3,2,1]
print(mergeSort(arr, 0, len(arr)-1))

File: Sorting1/MergeSort.java
public class MergeSort{

	
	public static int[] divide(int[] arr, int start, int end){
		if(start>=end) return arr;

		int mid = start + (end - start) / 2;
		arr = divide(arr, start, mid);
		arr = divide(arr, mid+1, end);
		return combine(arr, start, mid, end);
	}

	public static int[] combine(int[] arr, int start, int mid, int end){
		int i = start, j = mid+1, k = 0;
		int[] aux = new int[arr.length];


		while(i<=mid && j<=end)
			aux[k++] = arr[i] <=arr[j] ? arr[i++] : arr[j++];
		
		while(i<=mid)
			aux[k++] = arr[i++];

		while(j<=end)
			aux[k++] = arr[j++];	

		for( i=0;i<k;i++)
			arr[start + i] = aux[i];
		
		return arr;
	}


	public static void main(String[] args) {
		int[] arr  = {5,4,3,2,1};
		arr = divide(arr, 0, arr.length - 1);

		for(int i: arr){
			System.out.print(i+" ");
		}
	}
}